src/AspNet.Security.OAuth.Airtable/AirtableAuthenticationHandler.cs

```diff
@@ -0,0 +1,137 @@
+        [NotNull] ILoggerFactory logger,
+        [NotNull] UrlEncoder encoder)
+        : base(options, logger, encoder)
+    {
+    }
+
+    protected override async Task<AuthenticationTicket> CreateTicketAsync(
+        [NotNull] ClaimsIdentity identity,
+        [NotNull] AuthenticationProperties properties,
+        [NotNull] OAuthTokenResponse tokens)
+    {
+        using var request = new HttpRequestMessage(HttpMethod.Get, Options.UserInformationEndpoint);
+        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+        using var response =
+            await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.UserProfileErrorAsync(Logger, response, Context.RequestAborted);
+            throw new HttpRequestException("An error occurred while retrieving the user profile.");
+        }
+
+        await using var stream = await response.Content.ReadAsStreamAsync(Context.RequestAborted);
+        using var payload = await JsonDocument.ParseAsync(stream, cancellationToken: Context.RequestAborted);
+
+        var principal = new ClaimsPrincipal(identity);
+        var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
+        context.RunClaimActions();
+
+        await Events.CreatingTicket(context);
+        return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
+    }
+
+    protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull]OAuthCodeExchangeContext context)
+    {
+        var tokenRequestParameters = new Dictionary<string, string>
+        {
+            { "client_id", Options.ClientId },
+            { "redirect_uri", context.RedirectUri },
+            { "client_secret", Options.ClientSecret },
+            { "code", context.Code },
+            { "grant_type", "authorization_code" }
+        };
+
+        // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+        if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier))
+        {
+            tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+            context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+        }
+
+        using var requestMessage = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+        requestMessage.Headers.Accept.Add(
+            new MediaTypeWithQualityHeaderValue(MediaTypeNames.Application.FormUrlEncoded));
+        requestMessage.Content = new FormUrlEncodedContent(tokenRequestParameters);
+        var credentials = Convert.ToBase64String(Encoding.ASCII.GetBytes(
+            $"{Options.ClientId}:{Options.ClientSecret}"));
```

Comment:
```suggestion
        requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue(MediaTypeNames.Application.FormUrlEncoded));
        requestMessage.Content = new FormUrlEncodedContent(tokenRequestParameters);
        var credentials = Convert.ToBase64String(Encoding.ASCII.GetBytes($"{Options.ClientId}:{Options.ClientSecret}"));
```

Reply:
Note: it's not the standard flavor of the `client_secret_basic` client authentication, that expects the identifier and the secret to be percent-encoded first. Unless Airtable is known to implement the flawed version, we should probably use the standard version: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/cc5acd8e9838bd8a48f2403106b268d973bd3bc1/src/AspNet.Security.OAuth.Ebay/EbayAuthenticationHandler.cs#L92-L100

Reply:
Also, it seems weird to add formurl-encoded to the list of accepted content types when a JSON serializer is used to process the response. Does Airtable require doing that type of sorcery?

Reply:
You are right, thanks. It was my fault for adding formurl-encoded to the list of accepted content types.